src/compute/src/server.rs

```diff
@@ -96,19 +86,38 @@ pub async fn compute_node_serve(
     info!("> version: {} ({})", RW_VERSION, GIT_SHA);
 
     // Initialize all the configs
-    let storage_opts = Arc::new(StorageOpts::from(&config));
     let stream_config = Arc::new(config.streaming.clone());
     let batch_config = Arc::new(config.batch.clone());
 
     // Register to the cluster. We're not ready to serve until activate is called.
-    let meta_client = MetaClient::register_new(
+    let (meta_client, system_params) = MetaClient::register_new(
         &opts.meta_address,
         WorkerType::ComputeNode,
         &advertise_addr,
         opts.parallelism,
     )
     .await
     .unwrap();
+    let storage_opts = Arc::new(StorageOpts::from((&config, &system_params)));
+
+    // TODO(zhidong): Only read from system params in v0.1.18.
+    let state_store_url = system_params.state_store.as_ref().unwrap();
+    let state_store_url = if state_store_url.is_empty() {
+        if let Some(s) = opts.state_store.as_ref() {
+            s
+        } else {
+            panic!("State store url is neither specified from CLI args nor on meta node");
+        }
+    } else {
+        state_store_url
+    };
```

Comment:
Just in case there're more occurrences for this snippet, I suggest we add an extension trait for the `SystemParams` struct and make sure all accessing is from this. It will be better if we proxy all accessing and make the fields `private` somewhere.

Reply:
How to make generated fields private 🥵

Reply:
Fixed, check `SystemParamsReader`

https://github.com/risingwavelabs/risingwave/blob/9a665fbabe30f0316e42b2229b12ccd2a63ab9e7/src/rpc_client/src/meta_client.rs#L910-L970

Reply:
LGTM 🥵